### ISSUE TITLE
Vector Crashes during Loadtests when allowing backpressure

### ISSUE THREAD

**Issue by dygfloyd** [vectordotdev/vector#19184]
In Vector, we have a pipeline in place that has Datadog and Newrelic as a sinks. We are trying to allow backpressure only when sending to Datadog. 

Whenever we change:

`buffer.when_full = "drop_newest"`

...to...

`buffer.when_full = "block"`
 
...in both our logs and metrics Datadog sink, the vector instances start to crash during our loadtests. Everything else runs fine. When we loadtest with "drop_newest", it holds up fine. It's just when we switch it to allow backpressure, those instances of Vector memory fills to capacity, 100% CPU and crashes. Since they are hosted in k8s, the pods restart but then the cycle starts over again.

Here is an example of one of our metric sinks:

```
[sinks.sink_datadog_metrics_5]
type = "datadog_metrics"
inputs = ["fanout_del_5"]
default_api_key = "${DD_API_KEY}"
batch.timeout_secs = 5
batch.max_events = 2000
buffer.max_events = 200000
buffer.when_full = "drop_newest"
request.concurrency = "adaptive"
```

 Thanks!

**Comment by dsmith3197** [comment]
Could you share your full Vector configuration and the version of Vector you are running?

**Comment by dygfloyd** [comment]
I don't think I could post all of them. Our config is VERY big. (176 files) Is there a specific piece I could share? (ie: The configs for our datadog sinks) I'm checking to see if I might be able to send the config directly to a member of the vector team vs posting it on a public repo. That might also be an option.

**Comment by bruceg** [comment]
The most interesting bits would be `fanout_del_5` (presumably a transform) and all of the components feeding in to it. However, this does sound like an issue with buffers that we resolved in current versions, so knowing the version of Vector in use would be informative.

**Comment by dsmith3197** [comment]
@dygfloyd and I discussed this offline. Copying the gist of the resolution below for prosperity.

This behavior was experienced on Vector v0.30.2, and there are multiple factors at play that could cause the service to OOM once the buffer setting is changed.

1. There was a performance bottleneck in the Datadog Metrics sink that has been resolved since v0.30.2 ([pr](https://github.com/vectordotdev/vector/pull/18759)). This caused the Datadog Metrics sinks to apply significant back pressure upstream.
1. The configuration is large (>150 components). When the Datadog Metric sinks apply back pressure, this causes all of their upstream components to apply back pressure and so on, which causes a buffer to fill for each of those components. Each transform has a memory buffer with a capacity of `500` events associated with it. Each source has a memory buffer with a capacity of `1000 * num_cpus` associated with it. And finally, the buffer for each sink is configurable by the user. Given the size of the config, this can cause a significant number of events to buffer when back pressure is applied, depending upon the structure of the topology.
1. The pods were configured with a memory limit of 2gb. My best guess is that the memory limit was hit and the service OOM’d before all of the component in-memory buffers reached capacity.

Going forward, the best way to resolve this issue is to upgrade to the latest version of Vector and to increase the pods' memory limits.
